build/repo.targets

```diff
@@ -1,56 +1,195 @@
+
+    <PrepareDependsOn>$(PrepareDependsOn);FilterRepositories</PrepareDependsOn>
+    <CleanDependsOn>$(CleanDependsOn);CleanUniverseArtifacts</CleanDependsOn>
+    <BuildDependsOn>$(BuildDependsOn);CloneRepositories;BuildRepositories</BuildDependsOn>
   </PropertyGroup>
 
-  <!-- workaround. Using Sake as the intermediate means this property doesn't flow into repo builds. -->
-  <Target Name="SetBuildNumber" Condition="'$(BuildNumber)' != ''">
-    <SetEnvironmentVariable Variable="BuildNumber" Value="$(BuildNumber)" />
+  <Import
+    Project="$(_RepositoriesWithCommitsFile)"
+    Condition="Exists('$(_RepositoriesWithCommitsFile)')" />
+  <Import
+    Project="Repositories.props"
+    Condition="!Exists('$(_RepositoriesWithCommitsFile)')" />
+  <Import
+    Project="$(_CommitsFile)"
+    Condition="Exists('$(_CommitsFile)')" />
+
+  <Target Name="CleanUniverseArtifacts">
+    <RemoveDir Directories="obj" Condition="Exists('obj')" />
+    <RemoveDir Directories="$(_CloneRepositoryRoot)" Condition="Exists('$(_CloneRepositoryRoot)')" />
+    <RemoveDir Directories="$(_ArtifactsPackagesDirectory)" Condition="Exists('$(_ArtifactsPackagesDirectory)')" />
   </Target>
 
-  <Target Name="VerifyAll">
-    <PropertyGroup>
-      <_SakeTargets>$(_SakeTargets):verify-all</_SakeTargets>
-    </PropertyGroup>
-    <MSBuild Projects="$(MSBuildProjectFullPath)"
-      Targets="Sake"
-      Properties="SakeTargets=$(_SakeTargets);RepositoryRoot=$(RepositoryRoot)" />
+  <Target Name="FilterRepositories">
+    <ItemGroup Condition="'$(KOREBUILD_REPOSITORY_INCLUDE)'!=''">
+      <_RepositoriesToInclude Include="$(KOREBUILD_REPOSITORY_INCLUDE)" />
+      <Repository
+        Remove="@(Repository)"
+        Condition="'@(Repository)'!='@(_RepositoriesToInclude)' AND '%(Identity)'!=''" />
+    </ItemGroup>
+
+    <ItemGroup Condition="'$(KOREBUILD_REPOSITORY_EXCLUDE)'!=''">
+      <RepositoriesToExclude Include="$(KOREBUILD_REPOSITORY_EXCLUDE)" />
+      <Repository Remove="@(RepositoriesToExclude)" />
+    </ItemGroup>
+
+    <Warning Text="KOREBUILD_REPOSITORY_EXCLUDE AND KOREBUILD_REPOSITORY_INCLUDE are specified."
+      Condition="'$(KOREBUILD_REPOSITORY_INCLUDE)' != '' AND '$(KOREBUILD_REPOSITORY_EXCLUDE)' != ''" />
+
+    <Message Text="%(Repository.CloneUrl)" />
   </Target>
 
-  <Target Name="CIBuild" DependsOnTargets="SetBuildNumber">
+  <Target Name="CloneRepositories">
+    <ItemGroup>
+      <_CloneRepository Include="$(MSBuildProjectFullPath)">
+        <AdditionalProperties>
+          CloneRepository=%(Repository.Identity);
+          CloneUrl=%(Repository.CloneUrl);
+          CloneBranch=%(Repository.Branch);
+          CloneRepositoryCommit=%(Repository.Commit);
+          ShallowClone=true;
+          UseGateBranch=$(UseGateBranch)
+        </AdditionalProperties>
+      </_CloneRepository>
+    </ItemGroup>
+
+    <RemoveDir Directories="$(_CloneRepositoryRoot)" Condition="Exists('$(_CloneRepositoryRoot)')" />
```

Comment:
Can we not do this? When doing cross-repo work, I work inside the Universe folder. It's nice to repeatedly run universe commands without it deleting all my work.